docs/features/compilation-from-portable-pdb.md

```diff
@@ -0,0 +1,90 @@
+
+We will also store the MVID of a reference since it's a GUID that represents the symbol. This is to future proof the information we will need to look up references.
+
+At this time, we'll only include external references for the compilation. Any other references may be added later in a separate blob.
+
+## PDB Format Additions
+
+#### Metadata References
+
+Metadata references can be easily stored as binary. The binary encoding will be as follows (order matters):
+
+Name: A UTF-8 string (null terminated)
+Aliases: UTF-8 Comma (,) separate list of aliases (null terminated). May be empty
+MetadataImageKind: byte value representing Microsoft.CodeAnalysis.MetadataImageKind
+EmbedInteropTypes: boolean
+Timestamp: 4 byte integer
+File Size: 4 byte integer
+MVID: 16 byte integer (GUID)
+
+#### Compiler Flag Key Value Pairs
+
+The remaining values will be stored as key value pairs in the pdb. The storage format will be UTF8 encoded key value pairs that are null terminated. Order is not guaranteed. Any values left out can be assumed to be the default for the type. Keys may be different for Visual Basic and CSharp. They are serialized to reflect the command line arguments representing the same values
+
+Example: 
+
+
+`compilerversion\01.2.3\0sourceencoding\0utf-8\0checked\01\0unsafe\01\0langversion\0latest\0nullable\0Enable`
+
+##### Options For CSharp
+
+See [compiler options](https://docs.microsoft.com/en-us/dotnet/csharp/language-reference/compiler-options/listed-alphabetically) documentation
+
+1. "compilerversion"
+2. "[codepage](https://docs.microsoft.com/en-us/dotnet/csharp/language-reference/compiler-options/codepage-compiler-option)"
+3. "[checked](https://docs.microsoft.com/en-us/dotnet/csharp/language-reference/compiler-options/checked-compiler-option)"
+4. "[unsafe](https://docs.microsoft.com/en-us/dotnet/csharp/language-reference/compiler-options/unsafe-compiler-option)"
+5. "[langversion](https://docs.microsoft.com/en-us/dotnet/csharp/language-reference/compiler-options/langversion-compiler-option)"
+6. "[nullable](https://docs.microsoft.com/en-us/dotnet/csharp/nullable-references)"
+7. "[define](https://docs.microsoft.com/en-us/dotnet/csharp/language-reference/compiler-options/define-compiler-option)"
+
+##### Options For Visual Basic
+
+See [compiler options](https://docs.microsoft.com/en-us/dotnet/visual-basic/reference/command-line-compiler/compiler-options-listed-alphabetically) documentation
+
+1. "compilerversion"
+2. "[removeintchecks](https://docs.microsoft.com/en-us/dotnet/visual-basic/reference/command-line-compiler/removeintchecks)"
+3. "[langversion](https://docs.microsoft.com/en-us/dotnet/visual-basic/reference/command-line-compiler/langversion)"
+4. "[define](https://docs.microsoft.com/en-us/dotnet/visual-basic/reference/command-line-compiler/define)"
+5. "[optionstrict](https://docs.microsoft.com/en-us/dotnet/visual-basic/reference/command-line-compiler/optionstrict)"
```

Comment:
>[optionstrict] [](start = 4, length = 14)

What drives the decision whether an option should be included or excluded?

Reply:
The goal is to include enough information in the PDB so that a consumer could fully reconstruct the original compilation based on the information stored in the PDB, the assembly metadata, symbol and source servers and any private keys used to build the assembly.

Any compiler switches needed to meet that goal should be included

Reply:
> Any compiler switches needed to meet that goal should be included

I'd pose the question in the other direction: what switches can be excluded because they don't meaningfully impact the compilation? In general all switches impact the compilation, there are very few which do not.

Reply:
* appconfig
* baseaddress
* bugreport
* checksumalgorithm
* debug
* delaysign
* doc
* deterministic
* embed
* errorendlocation
* errorlog
* errorreport
* filealign
* fullpaths
* highentropyva
* incremental
* keycontainer
* keyfile
* link
* linkresource
* main
* moduleassemblyname
* modulename
* noconfig
* nologo
* nostdlib
* nowarn
* nowin32manifest
* optimize
* out
* parallel
* pathmap
* pdb
* platform
* preferreduilang
* publicsign
* recurse
* refout
* refonly
* reportanalyzer
* resource
* ruleset
* subsystemversion
* target
* utf8output
* version
* warn
* warnaserror
* win32icon
* win32manifest
* win32res

That's the list of all compiler switches (that I could find) that I believe are not needed because they are either implicitly or explicitly available through pdb information already, or are not needed to directly reconstruct the compilation in such a way that we could run analyzers on the input to the compilation and be confident that it represents the code and settings used to create the PE file.

Reply:
* checksumalgorithm

Seems feasible that analyzers would want to audit this. Consider that security analyzers frequently ban older check sum algorithms. 

* debug
* optimize

Where is this already stored that we could reconstruct it?

* link

This is essentially equivalent to `/reference` why would we exclude it?

* main

This meaningfully impacts the compilation, think it needs to be included. 

* moduleassemblyname
* modulename

There are lots of analyzers out there which focus on naming. Why would we exclude any name based arguments?

* nostdlib

This meaningfully impacts compilation. 

Overall I think we need to keep the full list of options in this document including the list of options we explicitly decided to not include. That way future developers have a baseline to consider. Lacking the full set of options in this document a developer can't tell for a given option if we considered and rejected or just didn't consider an option at all.

Reply:
- `checksumalgorithm` is already stored in the PDB for each source file. 
- optimization level should be included
- `link` is already included since it's represented by a metadata reference with EmbedInteropTypes=true. 
- `main` is already included in the PDB (the entry point token). 
- `moduleassemblyname`, `modulename` are in the metadata
- `nostdlib` is not needed as we have encode exact set of all references.

> Overall I think we need to keep the full list of options in this document including the list of options we explicitly decided to not include. That way future developers have a baseline to consider. Lacking the full set of options in this document a developer can't tell for a given option if we considered and rejected or just didn't consider an option at all.

+1

Reply:
I'll change documentation to include this full list, maybe with notes about why they aren't included

Reply:
Updated for VBC and CSC, let me know if there needs to be more information. Most of it is just the list of things missing, with a few noted that they already have information in the pdb

Reply:
Question: why not include all switches always?

Few reasons:

1. Easier to reason about. The compilers always take all the switches and store them, no need to parse, include/exclude specific ones. Less room for an error to slip by when introducing a new switch
2. These are baked into assets that will be around for a long time. We don't necessarily know all the possible ways that the values might be useful in the future. Having them in now future-proofs us.

What's the benefit of including/excluding specific switches?

Reply:
> What's the benefit of including/excluding specific switches?

One concrete reason is that it could break determinism. Consider that some switches in their raw form, such as source file paths, are mutated by other arguments to the compiler like `/pathMap`. Hence if we included all switches literally then it would break determinism because we'd have original paths vs. mapped paths in the PDB. 

So there does need to be some level of processing and exclusion here. Concretely including `pathMap` would in many ways defeat the purpose of having it 😦 

There are others which are questionable but it's hard to reason about because the goals of this are not clear. Is the goal that we need enough information such that we could recreate the binary with byte for byte equivalence? Essentially do we want enough info to guarantee that we can do the following:

1. Enough options are stored such that given the same inputs we could produce the same assembly (byte for byte equivalence)?
1. Enough option information is stored that we could detect when inputs are not the same. Example is that storing the MVID of a DLL is a strong barrier that we are presented with the right input when re-creating the compilation.

Reply:
> Is the goal that we need enough information such that we could recreate the binary with byte for byte equivalence? 

Yes.

> Enough options are stored such that given the same inputs we could produce the same assembly (byte for byte equivalence)?

Yes.

> Enough option information is stored that we could detect when inputs are not the same. Example is that storing the MVID of a DLL is a strong barrier that we are presented with the right input when re-creating the compilation.

We need information to find the metadata references. We include enough information so that we can look them up on symbol server, e.g. (or a similar storage).

Reply:
> Question: why not include all switches always?

That doesn't work for multiple reasons. One is determinism that Jared mentioned. Others are that some info can't be represented in the PDB in the same way it's represented on command line. Examples include references, for which we need ids that we can use to look the files up on symbol server. Another one is resources that will need to be retrieved from the DLL as we don't want to duplicate the resources in the PDB. We also don't want to duplicate other information that's easily retrievable from the DLL (such as various PE header flags) since whenever you have the same info stored in multiple places there is a potential for mismatch.

Reply:
Finally, there is also the aspect of Compilation APIs that are public. Some command line switches are not even available to the Compilation APIs because they are only needed by the command line compiler to find DLLs, sources, etc (such as varioys `/lib`, `/nostdlib` etc). It'd be unnecessary to add public APIs to EmitOptions to pass them in as we don't need these to recreate the compilation. We already have the info in other forms.